File: ts/project/.sdk/tm/java/test/StructCorpusTest.java
package JAVAPACKAGE.sdktest;

// Drives the shared struct corpus (../.sdk/test/test.json, root key "struct")
// against the vendored struct utility, mirroring the go target's
// struct_utility_test.go coverage. Adapted from the voxgig struct java
// port's StructCorpusTest; here the corpus is a green-bar test — any
// failing entry fails the build. Categories absent from this SDK's
// test.json are skipped.

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.TreeMap;
import java.util.function.Function;

import JAVAPACKAGE.utility.struct.Struct;

@SuppressWarnings({"unchecked", "rawtypes"})
public class StructCorpusTest {

  private static final Map<String, StructRunner.Result> SCOREBOARD = new TreeMap<>();

  private static Object getp(Object in, String key) {
    if (in instanceof Map<?, ?> m) {
      return ((Map<String, Object>) m).get(key);
    }
    return null;
  }

  private static Object getpDef(Object in, String key, Object def) {
    if (in instanceof Map<?, ?> m && ((Map<String, Object>) m).containsKey(key)) {
      return ((Map<String, Object>) m).get(key);
    }
    return def;
  }

  @TestFactory
  Iterable<DynamicTest> corpus() {
    List<DynamicTest> tests = new ArrayList<>();

    // ===== minor (29 names) =====
    add(tests, "minor", "isnode", true, in -> Struct.isnode(in));
    add(tests, "minor", "ismap", true, in -> Struct.ismap(in));
    add(tests, "minor", "islist", true, in -> Struct.islist(in));
    add(tests, "minor", "iskey", false, in -> Struct.iskey(in));
    add(tests, "minor", "strkey", false, in -> Struct.strkey(in));
    add(tests, "minor", "isempty", false, in -> Struct.isempty(in));
    add(tests, "minor", "isfunc", true, in -> Struct.isfunc(in));
    add(tests, "minor", "getprop", false, in -> {
      Object val = getp(in, "val");
      Object key = getp(in, "key");
      Object alt = getpDef(in, "alt", Struct.UNDEF);
      return alt == Struct.UNDEF
          ? Struct.getprop(val, key)
          : Struct.getprop(val, key, alt);
    });
    add(tests, "minor", "getelem", false, in -> {
      Object val = getp(in, "val");
      Object key = getp(in, "key");
      Object alt = getpDef(in, "alt", Struct.UNDEF);
      return alt == Struct.UNDEF
          ? Struct.getelem(val, key)
          : Struct.getelem(val, key, alt);
    });
    add(tests, "minor", "clone", false, in -> Struct.clone(in));
    add(tests, "minor", "items", true, in -> Struct.items(in));
    add(tests, "minor", "keysof", true, in -> Struct.keysof(in));
    add(tests, "minor", "haskey", false, in -> Struct.haskey(getp(in, "src"), getp(in, "key")));
    add(tests, "minor", "setprop", true, in -> {
      Object parent = getpDef(in, "parent", Struct.UNDEF);
      Object key = getp(in, "key");
      Object val = getp(in, "val");
      return Struct.setprop(parent == Struct.UNDEF ? null : parent, key, val);
    });
    add(tests, "minor", "delprop", true, in -> {
      Object parent = getpDef(in, "parent", Struct.UNDEF);
      Object key = getp(in, "key");
      return Struct.delprop(parent == Struct.UNDEF ? null : parent, key);
    });
    add(tests, "minor", "stringify", false, in -> {
      // null:false keeps a JSON-null val as a real null (rendered "null"); an
      // absent val is UNDEF (rendered ""). Mirrors the canonical harness.
      Object val = getpDef(in, "val", Struct.UNDEF);
      Object max = getp(in, "max");
      Integer m = max instanceof Number n ? n.intValue() : null;
      return Struct.stringify(val, m);
    });
    add(tests, "minor", "jsonify", false, in -> {
      Object val = getp(in, "val");
      Object flags = getp(in, "flags");
      return Struct.jsonify(val, flags);
    });
    add(tests, "minor", "pathify", false, in -> {
      // null:false keeps a JSON-null path as a real null ("<unknown-path:null>")
      // and an absent path as UNDEF ("<unknown-path>"); null parts are dropped.
      Object path = getpDef(in, "path", Struct.UNDEF);
      Object from = getp(in, "from");
      Object to = getp(in, "to");
      return Struct.pathify(path, from, to);
    });
    add(tests, "minor", "escre", true, in -> Struct.escre(in));
    add(tests, "minor", "escurl", true, in -> Struct.escurl(in));
    add(tests, "minor", "join", false, in -> {
      Object val = getp(in, "val");
      Object sep = getp(in, "sep");
      Object url = getp(in, "url");
      return Struct.join(val, sep, url);
    });
    add(tests, "minor", "flatten", true, in -> {
      Object val = getp(in, "val");
      Object depth = getp(in, "depth");
      Integer d = depth instanceof Number n ? n.intValue() : null;
      return Struct.flatten(val, d == null ? 1 : d);
    });
    add(tests, "minor", "filter", true, in -> {
      Object val = getp(in, "val");
      String check = Objects.toString(getp(in, "check"), "");
      Struct.ItemCheck pred = "gt3".equals(check)
          ? (Struct.ItemCheck) item -> {
            Object v = item.get(1);
            return v instanceof Number n && n.doubleValue() > 3;
          }
          : (Struct.ItemCheck) item -> {
            Object v = item.get(1);
            return v instanceof Number n && n.doubleValue() < 3;
          };
      return Struct.filter(val, pred);
    });
    add(tests, "minor", "typename", true, in -> Struct.typename(in));
    add(tests, "minor", "typify", false, in -> Struct.typify(in));
    add(tests, "minor", "size", false, in -> Struct.size(in));
    add(tests, "minor", "slice", false, in -> {
      Object val = getp(in, "val");
      Object start = getp(in, "start");
      Object end = getp(in, "end");
      return Struct.slice(val, start, end);
    });
    add(tests, "minor", "pad", false, in -> {
      Object val = getp(in, "val");
      Object pad = getp(in, "pad");
      Object pc = getp(in, "char");
      return Struct.pad(val, pad, pc);
    });
    add(tests, "minor", "setpath", false, in -> {
      Object store = getp(in, "store");
      Object path = getp(in, "path");
      Object val = getp(in, "val");
      return Struct.setpath(store, path, val);
    });

    // ===== walk =====
    add(tests, "walk", "basic", true, in -> Struct.walk(in,
        (k, v, p, t) -> v instanceof String s ? s + "~" + String.join(".", t) : v));
    add(tests, "walk", "depth", false, in -> {
      Object src = getp(in, "src");
      Object md = getp(in, "maxdepth");
      int maxdepth = md instanceof Number n ? n.intValue() : 32;
      Object[] top = new Object[1];
      Object[] cur = new Object[1];
      Struct.WalkApply copy =
          (key, val, parent, path) -> {
            if (key == null || Struct.isnode(val)) {
              Object child =
                  Struct.islist(val) ? new ArrayList<Object>() : new LinkedHashMap<String, Object>();
              if (key == null) {
                top[0] = child;
                cur[0] = child;
              } else {
                Struct.setprop(cur[0], key, child);
                cur[0] = child;
              }
            } else {
              Struct.setprop(cur[0], key, val);
            }
            return val;
          };
      Struct.walk(src, copy, null, maxdepth);
      return top[0];
    });
    add(tests, "walk", "copy", true, in -> {
      Object[] cur = new Object[64];
      Struct.WalkApply walkcopy =
          (key, val, parent, path) -> {
            if (key == null) {
              cur[0] =
                  Struct.ismap(val)
                      ? new LinkedHashMap<String, Object>()
                      : Struct.islist(val) ? new ArrayList<Object>() : val;
              return val;
            }
            Object v = val;
            int i = path.size();
            if (Struct.isnode(v)) {
              v =
                  Struct.ismap(v)
                      ? new LinkedHashMap<String, Object>()
                      : new ArrayList<Object>();
              cur[i] = v;
            }
            Struct.setprop(cur[i - 1], key, v);
            return val;
          };
      Struct.walk(in, walkcopy);
      return cur[0];
    });

    // ===== merge =====
    add(tests, "merge", "cases", true, in -> Struct.merge(in));
    add(tests, "merge", "array", true, in -> Struct.merge(in));
    add(tests, "merge", "integrity", true, in -> Struct.merge(in));
    add(tests, "merge", "depth", true, in -> {
      Object val = getp(in, "val");
      Object depth = getp(in, "depth");
      int d = depth instanceof Number n ? n.intValue() : 32;
      return Struct.merge(val, d);
    });

    // ===== getpath =====
    add(tests, "getpath", "basic", true, in -> Struct.getpath(getp(in, "store"), getp(in, "path")));
    add(tests, "getpath", "relative", true, in -> {
      Map<String, Object> inj = new LinkedHashMap<>();
      if (in instanceof Map<?, ?> m) {
        Map<String, Object> mm = (Map<String, Object>) m;
        if (mm.containsKey("dparent")) inj.put("dparent", mm.get("dparent"));
        if (mm.containsKey("dpath")) inj.put("dpath", mm.get("dpath"));
        if (mm.containsKey("base")) inj.put("base", mm.get("base"));
      }
      return Struct.getpath(getp(in, "store"), getp(in, "path"), inj.isEmpty() ? null : inj);
    });
    add(tests, "getpath", "special", true, in -> {
      Object inj = getp(in, "inj");
      Map<String, Object> injMap = inj instanceof Map<?, ?> m ? (Map<String, Object>) m : null;
      return Struct.getpath(getp(in, "store"), getp(in, "path"), injMap);
    });

    // ===== inject =====
    // inject.string passes the nullModifier so a resolved JSON null (encoded by
    // the runner's fixJSON as "__NULL__") renders as the literal text "null".
    add(tests, "inject", "string", true, in -> {
      Map<String, Object> opts = new LinkedHashMap<>();
      opts.put("modify", StructRunner.NULL_MODIFIER);
      return Struct.inject(getp(in, "val"), getp(in, "store"), opts);
    });
    add(tests, "inject", "deep", true, in -> Struct.inject(getp(in, "val"), getp(in, "store")));

    // ===== transform =====
    add(tests, "transform", "paths", true, in -> Struct.transform(getp(in, "data"), getp(in, "spec")));
    add(tests, "transform", "cmds", true, in -> Struct.transform(getp(in, "data"), getp(in, "spec")));
    add(tests, "transform", "each", true, in -> Struct.transform(getp(in, "data"), getp(in, "spec")));
    add(tests, "transform", "pack", true, in -> Struct.transform(getp(in, "data"), getp(in, "spec")));
    add(tests, "transform", "modify", true, in -> {
      Map<String, Object> opts = new LinkedHashMap<>();
      // Match JS test guard: only mutate string leaves.
      opts.put(
          "modify",
          (Struct.Modify)
              (val, key, parent, inj, store) -> {
                if (key != null && parent instanceof Map<?, ?> m && val instanceof String s) {
                  ((Map<String, Object>) m).put(Objects.toString(key), "@" + s);
                }
              });
      return Struct.transform(getp(in, "data"), getp(in, "spec"), opts);
    });
    add(tests, "transform", "ref", true, in -> Struct.transform(getp(in, "data"), getp(in, "spec")));
    add(tests, "transform", "format", false, in -> Struct.transform(getp(in, "data"), getp(in, "spec")));
    add(tests, "transform", "apply", true, in -> {
      Map<String, Object> opts = new LinkedHashMap<>();
      Map<String, Object> extra = new LinkedHashMap<>();
      extra.put(
          "apply",
          (Function<Object, Object>) v -> v instanceof String s ? s.toUpperCase(Locale.ROOT) : v);
      opts.put("extra", extra);
      return Struct.transform(getp(in, "data"), getp(in, "spec"), opts);
    });

    // ===== validate =====
    add(tests, "validate", "basic", false, in -> Struct.validate(getp(in, "data"), getp(in, "spec")));
    add(tests, "validate", "child", true, in -> Struct.validate(getp(in, "data"), getp(in, "spec")));
    add(tests, "validate", "one", true, in -> Struct.validate(getp(in, "data"), getp(in, "spec")));
    add(tests, "validate", "exact", true, in -> Struct.validate(getp(in, "data"), getp(in, "spec")));
    add(tests, "validate", "invalid", false, in -> Struct.validate(getp(in, "data"), getp(in, "spec")));
    add(tests, "validate", "special", true, in -> {
      Map<String, Object> inj = null;
      if (in instanceof Map<?, ?> m) {
        Object injObj = ((Map<String, Object>) m).get("inj");
        if (injObj instanceof Map<?, ?> im) {
          inj = (Map<String, Object>) im;
        }
      }
      return Struct.validate(getp(in, "data"), getp(in, "spec"), inj);
    });

    // ===== select =====
    add(tests, "select", "basic", true, in -> Struct.select(getp(in, "obj"), getp(in, "query")));
    add(tests, "select", "operators", true, in -> Struct.select(getp(in, "obj"), getp(in, "query")));
    add(tests, "select", "edge", true, in -> Struct.select(getp(in, "obj"), getp(in, "query")));
    add(tests, "select", "alts", true, in -> Struct.select(getp(in, "obj"), getp(in, "query")));

    return tests;
  }

  private void add(
      List<DynamicTest> tests,
      String category,
      String name,
      boolean nullFlag,
      StructRunner.Subject subject) {
    tests.add(
        DynamicTest.dynamicTest(
            category + "-" + name,
            () -> {
              Map<String, Object> spec = StructRunner.getSpec(category, name);
              if (spec == null) {
                // Category/name absent from this SDK's corpus: skip.
                return;
              }
              StructRunner.Result r =
                  StructRunner.runsetflags(category + "." + name, spec, nullFlag, subject);
              SCOREBOARD.put(category + "." + name, r);
              if (!r.failures.isEmpty()) {
                throw new AssertionError(
                    category + "." + name + ": " + r.passed + "/" + r.total
                        + " passed; failures:\n  " + String.join("\n  ", r.failures));
              }
            }));
  }

  @AfterAll
  static void printScoreboard() {
    int totalP = 0;
    int totalT = 0;
    StringBuilder banner = new StringBuilder();
    banner.append("\n========= STRUCT CORPUS SCOREBOARD =========\n");
    for (Map.Entry<String, StructRunner.Result> e : SCOREBOARD.entrySet()) {
      StructRunner.Result r = e.getValue();
      banner.append(String.format("  %-30s %4d / %4d%n", e.getKey(), r.passed, r.total));
      totalP += r.passed;
      totalT += r.total;
    }
    banner.append(String.format("  %-30s %4d / %4d%n", "TOTAL", totalP, totalT));
    banner.append("============================================\n");
    System.out.println(banner);
  }
}

File: ts/project/.sdk/tm/kotlin/test/StructCorpusTest.kt
package KOTLINPACKAGE.sdktest

// Drives the shared struct corpus (../.sdk/test/test.json, root key
// "struct") against the vendored Kotlin struct utility THROUGH the vendored
// omni runner (OmniResolver over test/vendor/omni) - the engine half of the
// retired StructRunner. Mirrors tm/java/test/StructCorpusTest.java and the
// go target's struct_utility_test.go coverage. A missing category or
// section FAILS (the old runner skipped it silently - a renamed fixture
// reported PASS while running zero assertions); a failing entry throws
// OmniError with the entry named.

import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assumptions
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestFactory

import java.util.function.Function

import voxgig.omni.Flags

import KOTLINPACKAGE.core.Helpers
import KOTLINPACKAGE.core.ProjectNameSDK
import KOTLINPACKAGE.utility.struct.Struct

@Suppress("UNCHECKED_CAST")
class StructCorpusTest {

  companion object {
    const val TEST_JSON_FILE = "../.sdk/test/test.json"

    private var RUN: OmniResolver.Run? = null

    @Synchronized
    fun structRun(): OmniResolver.Run {
      var r = RUN
      if (null == r) {
        r = OmniResolver
          .makeRunner(TEST_JSON_FILE, ProjectNameSDK.testSDK())
          .runner("struct", linkedMapOf<String, Any?>())
        assertNotNull(r.spec, "struct section not found in test.json")
        RUN = r
      }
      return r
    }
  }

  /** A subject in the struct-corpus shape: one argument in, one value out. */
  fun interface StructSubject {
    fun apply(input: Any?): Any?
  }

  private fun getp(input: Any?, key: String): Any? =
    if (input is Map<*, *>) (input as Map<String, Any?>)[key] else null

  private fun getpDef(input: Any?, key: String, def: Any?): Any? =
    if (input is Map<*, *> && (input as Map<String, Any?>).containsKey(key)) input[key] else def

  private fun hasKey(input: Any?, key: String): Boolean =
    input is Map<*, *> && (input as Map<String, Any?>).containsKey(key)

  @TestFactory
  fun corpus(): Iterable<DynamicTest> {
    val tests = mutableListOf<DynamicTest>()

    // ===== minor =====
    add(tests, "minor", "isnode", true) { Struct.isnode(it) }
    add(tests, "minor", "ismap", true) { Struct.ismap(it) }
    add(tests, "minor", "islist", true) { Struct.islist(it) }
    add(tests, "minor", "iskey", false) { Struct.iskey(it) }
    add(tests, "minor", "strkey", false) { Struct.strkey(it) }
    add(tests, "minor", "isempty", false) { Struct.isempty(it) }
    add(tests, "minor", "isfunc", true) { Struct.isfunc(it) }
    add(tests, "minor", "getprop", false) {
      val v = getp(it, "val")
      val k = getp(it, "key")
      val a = getpDef(it, "alt", Struct.UNDEF)
      if (a === Struct.UNDEF) Struct.getprop(v, k) else Struct.getprop(v, k, a)
    }
    add(tests, "minor", "getelem", false) {
      val v = getp(it, "val")
      val k = getp(it, "key")
      val a = getpDef(it, "alt", Struct.UNDEF)
      if (a === Struct.UNDEF) Struct.getelem(v, k) else Struct.getelem(v, k, a)
    }
    add(tests, "minor", "clone", false) { Struct.clone(it) }
    add(tests, "minor", "items", true) { Struct.items(it) }
    add(tests, "minor", "keysof", true) { Struct.keysof(it) }
    add(tests, "minor", "haskey", false) { Struct.haskey(getp(it, "src"), getp(it, "key")) }
    add(tests, "minor", "setprop", true) {
      val parent = getpDef(it, "parent", Struct.UNDEF)
      Struct.setprop(if (parent === Struct.UNDEF) null else parent, getp(it, "key"), getp(it, "val"))
    }
    add(tests, "minor", "delprop", true) {
      val parent = getpDef(it, "parent", Struct.UNDEF)
      Struct.delprop(if (parent === Struct.UNDEF) null else parent, getp(it, "key"))
    }
    add(tests, "minor", "stringify", false) {
      val v = if (hasKey(it, "val")) getp(it, "val") else Struct.UNDEF
      val maxlen = (getp(it, "max") as? Number)?.toInt()
      if (maxlen == null) Struct.stringify(v) else Struct.stringify(v, maxlen)
    }
    add(tests, "minor", "jsonify", false) {
      val v = if (hasKey(it, "val")) getp(it, "val") else Struct.UNDEF
      Struct.jsonify(v, getp(it, "flags"))
    }
    add(tests, "minor", "pathify", false) {
      val path = if (hasKey(it, "path")) getp(it, "path") else Struct.UNDEF
      Struct.pathify(path, getp(it, "from"), getp(it, "to"))
    }
    add(tests, "minor", "escre", true) { Struct.escre(it) }
    add(tests, "minor", "escurl", true) { Struct.escurl(it) }
    add(tests, "minor", "join", false) { Struct.join(getp(it, "val"), getp(it, "sep"), getp(it, "url")) }
    add(tests, "minor", "flatten", true) {
      Struct.flatten(getp(it, "val"), (getp(it, "depth") as? Number)?.toInt() ?: 1)
    }
    add(tests, "minor", "filter", true) {
      val v = getp(it, "val")
      if ("gt3" == getp(it, "check")) {
        Struct.filter(v) { item -> (item[1] as? Number)?.toDouble()?.let { d -> d > 3 } == true }
      } else {
        Struct.filter(v) { item -> (item[1] as? Number)?.toDouble()?.let { d -> d < 3 } == true }
      }
    }
    add(tests, "minor", "typename", true) { Struct.typename(it) }
    add(tests, "minor", "typify", false) { Struct.typify(it) }
    add(tests, "minor", "size", false) { Struct.size(it) }
    add(tests, "minor", "slice", false) { Struct.slice(getp(it, "val"), getp(it, "start"), getp(it, "end")) }
    add(tests, "minor", "pad", false) { Struct.pad(getp(it, "val"), getp(it, "pad"), getp(it, "char")) }
    add(tests, "minor", "setpath", false) { Struct.setpath(getp(it, "store"), getp(it, "path"), getp(it, "val")) }

    // ===== walk =====
    add(tests, "walk", "basic", true) {
      Struct.walk(it, Struct.WalkApply { _, v, _, p -> if (v is String) v + "~" + p.joinToString(".") else v })
    }
    add(tests, "walk", "depth", false) {
      val src = getp(it, "src")
      val maxdepth = (getp(it, "maxdepth") as? Number)?.toInt()
      val top = arrayOfNulls<Any>(1)
      val cur = arrayOfNulls<Any>(1)
      val copy = Struct.WalkApply { key, value, _, _ ->
        if (Struct.isnode(value)) {
          val child: Any? = if (Struct.islist(value)) mutableListOf<Any?>() else linkedMapOf<String, Any?>()
          if (key == null) {
            top[0] = child
            cur[0] = child
          } else {
            cur[0] = Struct.setprop(cur[0], key, child)
            cur[0] = child
          }
        } else if (key != null) {
          cur[0] = Struct.setprop(cur[0], key, value)
        }
        value
      }
      if (maxdepth == null) Struct.walk(src, copy) else Struct.walk(src, copy, null, maxdepth)
      top[0]
    }
    add(tests, "walk", "copy", true) {
      val cur = arrayOfNulls<Any>(64)
      val walkcopy = Struct.WalkApply { key, value, _, path ->
        if (key == null) {
          cur[0] = when {
            Struct.ismap(value) -> linkedMapOf<String, Any?>()
            Struct.islist(value) -> mutableListOf<Any?>()
            else -> value
          }
          value
        } else {
          var v: Any? = value
          val i = path.size
          if (Struct.isnode(v)) {
            v = if (Struct.ismap(v)) linkedMapOf<String, Any?>() else mutableListOf<Any?>()
            cur[i] = v
          }
          Struct.setprop(cur[i - 1], key, v)
          value
        }
      }
      Struct.walk(it, walkcopy)
      cur[0]
    }

    // ===== merge =====
    add(tests, "merge", "cases", true) { Struct.merge(it) }
    add(tests, "merge", "array", true) { Struct.merge(it) }
    add(tests, "merge", "integrity", true) { Struct.merge(it) }
    add(tests, "merge", "depth", true) {
      val v = getp(it, "val")
      val d = (getp(it, "depth") as? Number)?.toInt() ?: 32
      Struct.merge(v, d)
    }

    // ===== getpath =====
    add(tests, "getpath", "basic", true) { Struct.getpath(getp(it, "store"), getp(it, "path")) }
    add(tests, "getpath", "relative", true) {
      val inj: Struct.Injection? = (it as? Map<String, Any?>)?.let { m ->
        if (!m.containsKey("dparent") && !m.containsKey("dpath") && !m.containsKey("base")) {
          null
        } else {
          Struct.Injection(null, null).apply {
            if (m.containsKey("dparent")) dparent = m["dparent"]
            if (m.containsKey("dpath")) {
              when (val dp = m["dpath"]) {
                is List<*> -> dpath = dp.map { e -> e?.toString() ?: "" }.toMutableList()
                is String -> if (dp.isNotEmpty()) dpath = dp.split(".").toMutableList()
              }
            }
            if (m.containsKey("base") && m["base"] is String) base = m["base"] as String
          }
        }
      }
      Struct.getpath(getp(it, "store"), getp(it, "path"), inj)
    }
    add(tests, "getpath", "special", true) {
      val injMap = getp(it, "inj") as? Map<String, Any?>
      val inj = injMap?.let { im ->
        Struct.Injection(null, null).apply {
          if (im.containsKey("key")) key = im["key"]?.toString() ?: ""
          if (im.containsKey("dparent")) dparent = im["dparent"]
          if (im.containsKey("dpath")) {
            val dp = im["dpath"]
            if (dp is List<*>) dpath = dp.map { e -> e?.toString() ?: "" }.toMutableList()
          }
          if (im.containsKey("meta")) {
            val mm = im["meta"]
            if (mm is Map<*, *>) {
              meta = linkedMapOf<String, Any?>().also { for ((k, v) in mm) it[k.toString()] = v }
            }
          }
        }
      }
      Struct.getpath(getp(it, "store"), getp(it, "path"), inj)
    }

    // ===== inject =====
    // inject.string passes the null modifier so a resolved JSON null
    // (encoded by the runner as "__NULL__") renders as a real null / the
    // literal text "null" (was StructRunner.NULL_MODIFIER).
    add(tests, "inject", "string", true) {
      val inj = Struct.Injection(null, null)
      inj.modify = NULL_MODIFIER
      Struct.inject(getp(it, "val"), getp(it, "store"), inj)
    }
    add(tests, "inject", "deep", true) { Struct.inject(getp(it, "val"), getp(it, "store")) }

    // ===== transform =====
    add(tests, "transform", "paths", true) { Struct.transform(getp(it, "data"), getp(it, "spec")) }
    add(tests, "transform", "cmds", true) { Struct.transform(getp(it, "data"), getp(it, "spec")) }
    add(tests, "transform", "each", true) { Struct.transform(getp(it, "data"), getp(it, "spec")) }
    add(tests, "transform", "pack", true) { Struct.transform(getp(it, "data"), getp(it, "spec")) }
    add(tests, "transform", "modify", true) {
      val opts = linkedMapOf<String, Any?>(
        "modify" to Struct.Modify { v, k, parent, _, _ ->
          if (k != null && parent is MutableMap<*, *> && v is String) {
            (parent as MutableMap<String, Any?>)[k.toString()] = "@$v"
          }
        },
      )
      Struct.transform(getp(it, "data"), getp(it, "spec"), opts)
    }
    add(tests, "transform", "ref", true) { Struct.transform(getp(it, "data"), getp(it, "spec")) }
    add(tests, "transform", "format", false) { Struct.transform(getp(it, "data"), getp(it, "spec")) }
    add(tests, "transform", "apply", true) {
      val opts = linkedMapOf<String, Any?>(
        "extra" to linkedMapOf<String, Any?>(
          "apply" to Function<Any?, Any?> { v -> if (v is String) v.uppercase() else v },
        ),
      )
      Struct.transform(getp(it, "data"), getp(it, "spec"), opts)
    }

    // ===== validate =====
    add(tests, "validate", "basic", false) { Struct.validate(getp(it, "data"), getp(it, "spec")) }
    add(tests, "validate", "child", true) { Struct.validate(getp(it, "data"), getp(it, "spec")) }
    add(tests, "validate", "one", true) { Struct.validate(getp(it, "data"), getp(it, "spec")) }
    add(tests, "validate", "exact", true) { Struct.validate(getp(it, "data"), getp(it, "spec")) }
    add(tests, "validate", "invalid", false) { Struct.validate(getp(it, "data"), getp(it, "spec")) }
    add(tests, "validate", "special", true) {
      val inj = getp(it, "inj") as? Map<String, Any?>
      Struct.validate(getp(it, "data"), getp(it, "spec"), inj)
    }

    // ===== select =====
    add(tests, "select", "basic", true) { Struct.select(getp(it, "obj"), getp(it, "query")) }
    add(tests, "select", "operators", true) { Struct.select(getp(it, "obj"), getp(it, "query")) }
    add(tests, "select", "edge", true) { Struct.select(getp(it, "obj"), getp(it, "query")) }
    add(tests, "select", "alts", true) { Struct.select(getp(it, "obj"), getp(it, "query")) }

    return tests
  }

  // The struct.nullsem section: does a PRESENT key holding a JSON null read
  // as "no value"? Opt-in per target (create-sdkgen ships it; an older
  // project corpus may predate it - the abort below says so OUT LOUD, as a
  // skipped test, rather than passing vacuously). The vendored kotlin
  // struct answers the canonical way on every question, list form included
  // (measured; see docs/design/vendor-tag-rollout.md). All lanes run
  // {null: false}: without the flag the runner rewrites every null to
  // '__NULL__' and the section asserts nothing about null at all.
  @Test
  fun nullsem() {
    val run = structRun()
    val nullsem = Helpers.toMapAny(run.spec["nullsem"])
    if (nullsem == null) {
      Assumptions.abort<Any>(
        "corpus predates struct.nullsem - refresh .sdk/test/struct from create-sdkgen",
      )
      return
    }
    val flags = Flags(nulls = false)

    run.runsetflags(nullsem["getprop"], flags) { args ->
      val input = args.getOrNull(0)
      val alt = getpDef(input, "alt", Struct.UNDEF)
      if (alt === Struct.UNDEF) {
        Struct.getprop(getp(input, "val"), getp(input, "key"))
      } else {
        Struct.getprop(getp(input, "val"), getp(input, "key"), alt)
      }
    }

    run.runsetflags(nullsem["getelem"], flags) { args ->
      val input = args.getOrNull(0)
      val alt = getpDef(input, "alt", Struct.UNDEF)
      if (alt === Struct.UNDEF) {
        Struct.getelem(getp(input, "val"), getp(input, "key"))
      } else {
        Struct.getelem(getp(input, "val"), getp(input, "key"), alt)
      }
    }

    run.runsetflags(nullsem["getpath"], flags) { args ->
      Struct.getpath(getp(args.getOrNull(0), "store"), getp(args.getOrNull(0), "path"))
    }

    run.runsetflags(nullsem["haskey"], flags) { args ->
      Struct.haskey(getp(args.getOrNull(0), "src"), getp(args.getOrNull(0), "key"))
    }

    run.runsetflags(nullsem["keysof"], flags) { args ->
      Struct.keysof(args.getOrNull(0))
    }
  }

  private fun add(
    tests: MutableList<DynamicTest>,
    category: String,
    name: String,
    nullFlag: Boolean,
    subject: StructSubject,
  ) {
    tests.add(
      DynamicTest.dynamicTest("$category-$name") {
        val run = structRun()
        val cat = Helpers.toMapAny(run.spec[category])
        assertNotNull(cat, "struct corpus category missing: $category - check .sdk/test/struct/")
        val spec = Helpers.toMapAny(cat!![name])
        assertNotNull(spec, "struct corpus section missing: $category.$name - check .sdk/test/struct/")
        val set = spec!!["set"]
        assertTrue(
          set is List<*> && set.isNotEmpty(),
          "struct corpus section is EMPTY: $category.$name - zero cases would run",
        )
        run.runsetflags(spec, Flags(nulls = nullFlag, name = "$category.$name")) { args ->
          subject.apply(if (args.isNotEmpty()) args[0] else Struct.UNDEF)
        }
      },
    )
  }

  private val NULL_MODIFIER = Struct.Modify { value, key, parent, _, _ ->
    if (value is String) {
      val repl: Any? = when {
        value == OmniResolver.NULLMARK -> null
        value.contains(OmniResolver.NULLMARK) -> value.replace(OmniResolver.NULLMARK, "null")
        else -> return@Modify
      }
      when {
        parent is MutableMap<*, *> && key != null -> (parent as MutableMap<String, Any?>)[key.toString()] = repl
        parent is MutableList<*> && key is Number -> (parent as MutableList<Any?>)[key.toInt()] = repl
      }
    }
  }
}
